### DISCUSSION TITLE
Changing UploadedFile to not always write to disk

### DISCUSSION THREAD

**Discussion by refi64** [luckyframework/lucky#1640] [Feature requests & Ideas]
Right now `Lucky::UploadedFile` always saves the content to a temporary file before control is passed into the application. This has a few issues:

- There's no way to limit whether or not the file is downloaded based on its metadata / size.
- The size of uploaded files is always limited to the size of the temporary storage location.
- The file can't be directly streamed elsewhere (e.g. sent to a microservice or uploaded) without hitting the disk / tmpfs.

Now, here's my idea: `UploadedFile` doesn't actually save the file yet, the constructor just takes the given IO value and saves it. Then, a new method `.save` is added that will save it to disk and return the `File` instance, and another method `.content` / `.body` is added to return the IO stream. (`.path` would also have to be removed.) That way, if someone really needs it on disk, they can easily change their code, but anyone who can handle the direct stream doesn't need to save to disk first.

Thoughts?

**Comment by matthewmcgarvey** [comment]
@refi64 Could I ask for some examples of how other frameworks do what you're looking for. I think get what you're asking for, but I want to make sure we land on a good solution that makes sense to everyone involved.

**Reply by refi64** [reply]
Well, it turns out, I have absolutely *no* idea what I'm talking about! It turns out, pretty much *no* web frameworks support this, seemingly for the reason that, because multipart data comes in sequentially, you can't actually know what you're getting until you have it all.

However, it turns out a lot of frameworks *do* have an API similar to what I proposed, except for the complete opposite reason: *small* files are saved in memory, to avoid hitting the disk. So I was on the right track but for the wrong reason :sweat_smile: Some examples:

- [Django's `UploadedFile`](https://github.com/atom/github/issues/2032) lets you `read()` the contents of the file, with two subclasses (`MemoryUploadedFile`, `TemporaryUploadedFile`) handling the specific storage mechanism. It also mentions the behavior I described:
  > Together the `MemoryFileUploadHandler` and `TemporaryFileUploadHandler` provide Django’s default file upload behavior of reading small files into memory and large ones onto disk.
  Now, they also seem to let you choose to e.g. always do in-memory or always do on-disk, but I'm not sure how useful that type of control is.
- Flask uses [`werkzeug.datastructures.FileStorage`](https://werkzeug.palletsprojects.com/en/2.0.x/datastructures/#werkzeug.datastructures.FileStorage), which has a *very* similar API to what I had mentioned: `.stream` lets you stream the contents, and `.save(dst)` will save the file to the location `dst`.
- [FastAPI's `UploadFile`](https://fastapi.tiangolo.com/tutorial/request-files/) lets you treat is as a file-like object (akin to Crystal's IO), but if it's backed by a path on-disk, you cannot access it.
- Phoenix [uses `Plug.Upload`](https://hexdocs.pm/plug/Plug.Upload.html) which has a very similar API to Lucky's current `UploadedFile`.

Now, of course, this doesn't work for my use case of restricting the upload size...but most of the above frameworks seem to have support for that built in, so I'm guessing Lucky should do so as well? (That being said, I could also just stick it behind nginx and manage the upload size through there.)

**Reply by matthewmcgarvey** [reply]
Thanks for doing that. I makes it easier for me to conceptualize what your asking for if you can point to what others do 😄 

So to summarize things I've taken away from your comment:

- We can't know the size of file uploads until we have fully read the IO
- We provide a solution that's pretty standard across frameworks at the moment
- We have no built-in way to validate the file size of `UploadedFile`
- We have no way to "pipe" the IO without reading it

So I think if we say the first point is an understandable limitation, I think we could boil this down to two different updates to Lucky.

1. We could add a validation for file size (this would be an update to Avram, see [`Avram::Uploadable`](https://github.com/luckyframework/avram/blob/master/src/avram/uploadable.cr))
2. We update `Lucky::UploadedFile` api to not immediately create a tempfile with the IO data
a. I think it might make sense that `Lucky::UploadedFile` could be a fork in the road of sorts. You could provide two main methods `#tempfile` and `#raw`. The first would return an object with the same API we currently have, the other would just be the IO or the `HTTP::FormData::Part`.

Looking at this https://luckyframework.org/guides/handling-files/file-uploads I'm not sure how the new API would work with Avram, though.

@jwoertink Maybe you could weigh in?

**Reply by matthewmcgarvey** [reply]
Personally, I think I'm OK if you wanted to open an issue on Avram for adding the file size validation and starting that. I think the other might require a bit more discussion

**Reply by jwoertink** [reply]
We've been doing quite a bit of upload stuff. Not quite with the Lucky::UploadedFile, but @russ might be able to weigh in on things we could port over or not.

I'm with @matthewmcgarvey on the opening an issue for size validation. I think that sounds fine.

**Reply by refi64** [reply]
> Personally, I think I'm OK if you wanted to open an issue on Avram for adding the file size validation and starting that.

Well, from what I can tell, the other frameworks (well, at least Flask and Django) generally check the content size during the download, not afterwards, so it would still be Lucky issue :sweat_smile:

**Reply by jwoertink** [reply]
Maybe it could be two fold? SaveOperations that need a file should have some sort of validation method to ensure the file matches some criteria. So in that case, it would be an Avram issue. But we could maybe also add some sort of "operation-like" object in Lucky that could be some sort of base wrapper around uploading. I'm picturing something similar to [Carrierwave](https://github.com/carrierwaveuploader/carrierwave) uploader. Lucky wouldn't come with any sort of image processing, but maybe this could be a thing where if you had your own processor, you just include in your uploader class....

I'm still waking up here, so it's a super quarter-baked idea, but maybe that sparks a few other better ideas? 😬
